provides.py

```diff
@@ -59,15 +55,13 @@ def publish_url(self, vault_url, remote_binding=None):
         """
         for relation in self.relations:
             if remote_binding:
-                units = relation.units
-                if units:
-                    addr = units[0].received['ingress-address'] or \
-                        units[0].received['access_address']
-                    bound_cidr = resolve_network_cidr(
-                        network_get_primary_address(remote_binding)
-                    )
-                    if not (addr and is_address_in_network(bound_cidr, addr)):
-                        continue
+                binding_nets = set(network_get(remote_binding)
+                                   ['egress-subnets'])
+                relation_nets = set(network_get(self.endpoint_name,
+                                                relation.relation_id)
+                                    ['egress-subnets'])
+                if not (binding_nets & relation_nets):
+                    continue
```

Comment:
In essence, this is the LP#1895185 bug. When the charm calls publish_url it should always set vault_url. The bug is that there are a number of situations in the original (if not (addr and is_address_in_network(bound_cidr, addr)) where it does not get set. I do not want to repeat the same mistake.

Ultimately, this logic should be in the charm and not in the interface. publish_url should always set all the necessary variables for a functioning relationship.